File: client_leger/lib/services/game_replay_command_factory_service.dart
import 'package:client_leger/classes/game-replay-commands/end_game_game_replay_command.dart';
import 'package:client_leger/classes/game-replay-commands/game_replay_command.dart';
import 'package:client_leger/classes/game-replay-commands/game_start_game_replay_command.dart';
import 'package:client_leger/classes/game-replay-commands/guess_difference_game_replay_command.dart';
import 'package:client_leger/classes/game-replay-commands/message_game_replay_command.dart';
import 'package:client_leger/classes/game-replay-commands/use_hint_game_replay_command.dart';
import 'package:client_leger/interfaces/dto/replay_event.dto.dart';
import 'package:get/get.dart';

class GameReplayCommandFactoryService extends GetxController {
  static GameReplayCommandFactoryService get to => Get.find();

  List<GameReplayCommand> transformReplayEventsToCommands(
      List<BaseReplayEventDto> events) {
    List<GameReplayCommand> commands = [];

    for (BaseReplayEventDto event in events) {
      GameReplayCommand? command = _constructCommand(event);
      if (command != null) {
        commands.add(command);
      }
    }

    return commands;
  }

  GameReplayCommand? _constructCommand(BaseReplayEventDto event) {
    switch (event.event) {
      case ReplayEvent.gameStart:
        return GameStartGameReplayCommand(
            eventDto: event as GameStartReplayDto, time: event.time);
      case ReplayEvent.guessDifference:
        return GuessDifferenceGameReplayCommand(
            eventDto: event as GuessDifferenceReplayDto, time: event.time);
      case ReplayEvent.useHint:
        return UseHintGameReplayCommand(
            eventDto: event as UseHintReplayDto, time: event.time);
      case ReplayEvent.message:
        return MessageGameReplayCommand(
            eventDto: event as MessageReplayDto, time: event.time);
      case ReplayEvent.endGame:
        return EndGameGameReplayCommand(
            eventDto: event as EndGameReplayDto, time: event.time);
      default:
        return null;
    }
  }
}

File: client/src/app/services/game-replay-command-factory/game-replay-command-factory.service.ts
import { Injectable } from '@angular/core';
import { EndGameGameReplayCommand } from '@app/classes/end-game-game-replay-command/end-game-game-replay-command';
import { GameStartGameReplayCommand } from '@app/classes/game-start-game-replay-command/game-start-game-replay-command';
import { GuessDifferenceGameReplayCommand } from '@app/classes/guess-difference-game-replay-command/guess-difference-game-replay-command';
import { MessageGameReplayCommand } from '@app/classes/message-game-replay-command/message-game-replay-command';
import { UseHintGameReplayCommand } from '@app/classes/use-hint-game-replay-command/use-hint-game-replay-command';
import { GameReplayCommand } from '@app/interfaces/game-replay-command';
import { ClassicModeService } from '@app/services/classic-mode/classic-mode.service';
import { GameSessionEvent } from '@common/model/events/game-session.events';
import { ReplayEventDto } from '@common/model/events/replay.events';

@Injectable({
    providedIn: 'root',
})
export class GameReplayCommandFactoryService {
    transformReplayEventsToCommands(events: ReplayEventDto[], classicModeService: ClassicModeService): GameReplayCommand[] {
        return events.reduce((constructedCommands: GameReplayCommand[], replayEvent: ReplayEventDto) => {
            const gameReplayCommand = this.constructEvent(replayEvent, classicModeService);

            if (gameReplayCommand) {
                constructedCommands.push(gameReplayCommand);
            }

            return constructedCommands;
        }, []);
    }

    private constructEvent(replayEventDto: ReplayEventDto, classicGameService: ClassicModeService): GameReplayCommand | null {
        switch (replayEventDto.event) {
            case GameSessionEvent.GameStart:
                return new GameStartGameReplayCommand(classicGameService, replayEventDto);
            case GameSessionEvent.GuessDifference:
                return new GuessDifferenceGameReplayCommand(classicGameService, replayEventDto);
            case GameSessionEvent.UseHint:
                return new UseHintGameReplayCommand(classicGameService, replayEventDto);
            case GameSessionEvent.Message:
                return new MessageGameReplayCommand(classicGameService, replayEventDto);
            case GameSessionEvent.EndGame:
                return new EndGameGameReplayCommand(classicGameService, replayEventDto);
            default:
                return null;
        }
    }
}
